### ISSUE TITLE
Problem with exporting and importing AsymetricKey 

### ISSUE THREAD

**Issue by LKuzma** [AArnott/PCLCrypto#15]
I'm trying to do something like this:
```c#
var asymKey = WinRTCrypto.AsymmetricKeyAlgorithmProvider.OpenAlgorithm (AsymmetricAlgorithm.RsaOaepSha256).CreateKeyPair(2048);
var exported = asymKey.Export();

foreach (var item in exported)
    System.Diagnostics.Debug.Write(item + " ");

var imported = WinRTCrypto.AsymmetricKeyAlgorithmProvider.OpenAlgorithm(AsymmetricAlgorithm.RsaOaepSha256).ImportKeyPair(exported);

foreach (var item in imported.Export())
    System.Diagnostics.Debug.Write(item + " ");
```

I am getting different values from ICryptographicKey.KeySize before invoking Export and after invoking Import.
My project is Xamarin.Android min. API 15 target API 21
I noticed that the problem appears on the android 4.0.3 and 5.0, in 4.1.1 and 4.4.1 is working fine.

**Comment by AArnott** [comment]
It might be helpful if you include the output of this code you have, and any other detail you have.

**Comment by LKuzma** [comment]
I use:
PCLCrypto.1.0.0.15071
Validation.2.0.6.15003
I started a new project and build it in VS and Xamarin with the same result.
sample code:
```c#
var asymKey = WinRTCrypto.AsymmetricKeyAlgorithmProvider.OpenAlgorithm (AsymmetricAlgorithm.RsaOaepSha256).CreateKeyPair (2048);
System.Diagnostics.Debug.WriteLine("KeySize before export: " + asymKey.KeySize);
var exported = asymKey.Export();
            
System.Diagnostics.Debug.WriteLine("exported data:");
foreach (var item in exported)
    sbuild.Append(item + " ");
System.Diagnostics.Debug.WriteLine(sbuild.ToString());
sbuild.Clear();
			
var imported = WinRTCrypto.AsymmetricKeyAlgorithmProvider.OpenAlgorithm(AsymmetricAlgorithm.RsaOaepSha256).ImportKeyPair(exported);
System.Diagnostics.Debug.WriteLine("KeySize after import: " + imported.KeySize);

System.Diagnostics.Debug.WriteLine("imported data:");
foreach (var item in imported.Export ())
    sbuild.Append(item + " ");
System.Diagnostics.Debug.WriteLine(sbuild.ToString());
```
result:
```
KeySize before export: 2048
exported data:
48 130 4 205 2 1 0 48 13 6 9 42 134 72 134 247 13 1 1 1 5 0 4 130 4 168 48 130 4 164 2 1 0 2 130 1 1 0 191 71 186 24 162 139 175 235 105 172 194 76 41 148 158 72 172 25 56 218 38 255 105 166 19 253 240 95 253 174 143 117 52 224 188 253 46 176 246 38 146 226 193 47 30 183 35 107 56 39 108 254 106 25 110 231 28 95 215 77 56 125 184 24 202 60 105 197 155 37 142 203 115 192 61 108 17 237 76 164 45 199 240 140 22 161 246 85 72 110 205 79 197 49 102 66 5 207 61 137 202 122 201 120 18 52 253 221 96 91 54 16 206 150 252 25 46 243 0 220 152 3 6 0 242 147 46 204 209 146 242 113 11 70 154 10 52 166 42 183 240 60 21 107 192 162 252 5 144 150 90 34 136 22 225 47 229 29 46 154 212 109 74 238 107 78 241 224 76 58 71 186 89 237 224 187 239 172 128 180 50 193 50 245 72 184 2 194 38 76 44 218 50 35 217 193 207 116 243 50 150 20 162 180 214 51 192 205 223 114 161 65 250 225 101 215 220 171 54 187 242 169 68 231 120 0 124 24 176 175 204 70 24 156 23 151 246 141 88 119 247 39 85 154 101 135 224 156 192 2 181 101 151 147 157 177 2 3 1 0 1 2 130 1 0 114 165 218 60 3 201 56 3 33 14 196 82 132 37 169 13 163 225 102 59 19 94 201 131 26 190 125 1 251 250 81 208 13 134 154 160 34 9 61 48 119 159 227 236 139 59 119 198 63 29 26 110 15 71 249 58 26 190 179 60 124 156 149 232 125 230 225 250 231 248 5 243 154 8 98 136 80 226 107 170 136 42 43 218 187 130 14 52 165 186 139 226 193 45 207 189 200 223 163 97 224 115 226 39 206 185 230 99 118 234 148 168 84 59 25 21 59 23 185 175 126 137 105 98 206 2 59 236 204 35 189 211 40 81 207 75 76 131 53 52 36 83 28 135 32 113 29 249 199 92 50 21 127 15 40 231 56 243 121 168 199 28 122 190 103 75 213 192 183 190 9 215 186 114 43 244 21 128 170 130 206 75 209 113 194 39 195 193 12 3 41 198 210 111 177 190 35 221 199 155 4 220 26 111 72 170 152 123 203 32 100 255 153 17 110 231 255 108 227 208 242 57 115 169 25 148 1 232 119 178 176 112 254 22 107 73 229 112 50 169 163 16 0 241 251 107 246 139 10 71 200 70 179 36 16 241 2 129 129 0 232 141 182 101 190 214 116 113 233 228 237 199 74 207 244 145 6 28 235 140 85 151 53 196 87 77 64 13 50 156 31 196 110 180 6 137 115 254 2 224 74 227 126 13 214 142 87 226 223 51 62 3 178 5 255 119 0 39 147 82 122 155 21 141 125 128 236 133 131 162 191 219 98 227 41 216 62 175 229 226 32 137 220 56 103 25 255 177 180 102 127 38 166 222 212 119 117 122 156 152 251 169 21 52 229 65 211 52 140 210 88 92 99 39 76 86 81 228 123 164 44 234 237 34 218 179 31 93 2 129 129 0 210 144 187 249 242 241 225 50 93 120 1 185 19 218 143 245 8 49 133 110 81 141 122 234 234 20 45 231 68 180 138 24 56 214 164 17 29 192 178 32 83 185 77 189 36 158 253 73 191 213 7 65 172 108 181 216 113 186 199 28 211 132 68 120 86 37 193 218 26 218 14 161 127 185 41 231 104 223 100 221 228 144 104 206 139 101 204 100 204 69 235 213 11 232 57 43 37 236 98 115 135 118 190 114 89 38 61 2 69 44 217 8 49 0 34 71 96 230 61 178 218 245 104 63 56 235 86 101 2 129 129 0 161 82 133 96 105 149 109 108 4 124 57 242 209 84 23 109 111 87 42 147 204 96 59 253 59 46 172 171 59 27 144 241 252 117 200 63 28 6 238 201 215 73 159 13 76 116 172 69 6 223 14 59 20 9 19 33 227 42 145 9 91 98 47 254 90 60 208 87 2 109 41 56 176 141 179 94 126 134 25 43 215 180 170 223 57 174 144 113 31 27 39 67 145 232 236 108 254 67 19 85 16 231 71 125 205 55 239 162 249 185 133 112 121 156 46 89 135 47 192 17 176 211 59 179 189 17 27 69 2 129 128 155 98 249 133 35 86 130 20 51 203 186 161 196 215 246 130 33 112 194 1 190 39 4 253 84 142 93 173 156 19 11 17 130 232 74 156 226 242 239 192 81 104 158 200 164 188 187 55 240 91 153 134 74 98 210 243 14 66 130 107 112 95 220 83 93 103 113 157 173 81 27 152 236 115 77 76 27 37 55 255 83 139 137 252 89 252 94 144 71 250 149 101 91 235 28 61 14 56 90 18 221 117 176 103 28 8 206 222 221 3 146 137 171 247 104 84 122 90 142 79 109 171 202 41 85 172 189 205 2 129 129 0 77 210 90 188 159 105 138 233 113 142 192 153 120 129 140 82 66 148 67 237 85 220 232 248 240 217 135 10 86 126 5 233 8 5 179 130 164 70 182 77 227 14 207 158 58 144 44 211 251 160 136 50 129 194 118 121 100 126 138 136 35 212 92 187 152 55 117 133 47 251 199 86 197 130 191 188 40 0 154 205 226 71 106 57 132 60 254 125 84 30 157 42 48 52 15 39 65 125 106 105 171 152 166 155 235 19 139 115 144 16 125 83 137 51 87 96 32 19 73 229 201 98 207 114 106 160 70 74 160 13 48 11 6 3 85 29 15 49 4 3 2 0 16

KeySize after import: 2047
imported data:
48 130 4 205 2 1 0 48 13 6 9 42 134 72 134 247 13 1 1 1 5 0 4 130 4 168 48 130 4 164 2 1 0 2 130 1 1 0 191 71 186 24 162 139 175 235 105 172 194 76 41 148 158 72 172 25 56 218 38 255 105 166 19 253 240 95 253 174 143 117 52 224 188 253 46 176 246 38 146 226 193 47 30 183 35 107 56 39 108 254 106 25 110 231 28 95 215 77 56 125 184 24 202 60 105 197 155 37 142 203 115 192 61 108 17 237 76 164 45 199 240 140 22 161 246 85 72 110 205 79 197 49 102 66 5 207 61 137 202 122 201 120 18 52 253 221 96 91 54 16 206 150 252 25 46 243 0 220 152 3 6 0 242 147 46 204 209 146 242 113 11 70 154 10 52 166 42 183 240 60 21 107 192 162 252 5 144 150 90 34 136 22 225 47 229 29 46 154 212 109 74 238 107 78 241 224 76 58 71 186 89 237 224 187 239 172 128 180 50 193 50 245 72 184 2 194 38 76 44 218 50 35 217 193 207 116 243 50 150 20 162 180 214 51 192 205 223 114 161 65 250 225 101 215 220 171 54 187 242 169 68 231 120 0 124 24 176 175 204 70 24 156 23 151 246 141 88 119 247 39 85 154 101 135 224 156 192 2 181 101 151 147 157 177 2 3 1 0 1 2 130 1 0 114 165 218 60 3 201 56 3 33 14 196 82 132 37 169 13 163 225 102 59 19 94 201 131 26 190 125 1 251 250 81 208 13 134 154 160 34 9 61 48 119 159 227 236 139 59 119 198 63 29 26 110 15 71 249 58 26 190 179 60 124 156 149 232 125 230 225 250 231 248 5 243 154 8 98 136 80 226 107 170 136 42 43 218 187 130 14 52 165 186 139 226 193 45 207 189 200 223 163 97 224 115 226 39 206 185 230 99 118 234 148 168 84 59 25 21 59 23 185 175 126 137 105 98 206 2 59 236 204 35 189 211 40 81 207 75 76 131 53 52 36 83 28 135 32 113 29 249 199 92 50 21 127 15 40 231 56 243 121 168 199 28 122 190 103 75 213 192 183 190 9 215 186 114 43 244 21 128 170 130 206 75 209 113 194 39 195 193 12 3 41 198 210 111 177 190 35 221 199 155 4 220 26 111 72 170 152 123 203 32 100 255 153 17 110 231 255 108 227 208 242 57 115 169 25 148 1 232 119 178 176 112 254 22 107 73 229 112 50 169 163 16 0 241 251 107 246 139 10 71 200 70 179 36 16 241 2 129 129 0 232 141 182 101 190 214 116 113 233 228 237 199 74 207 244 145 6 28 235 140 85 151 53 196 87 77 64 13 50 156 31 196 110 180 6 137 115 254 2 224 74 227 126 13 214 142 87 226 223 51 62 3 178 5 255 119 0 39 147 82 122 155 21 141 125 128 236 133 131 162 191 219 98 227 41 216 62 175 229 226 32 137 220 56 103 25 255 177 180 102 127 38 166 222 212 119 117 122 156 152 251 169 21 52 229 65 211 52 140 210 88 92 99 39 76 86 81 228 123 164 44 234 237 34 218 179 31 93 2 129 129 0 210 144 187 249 242 241 225 50 93 120 1 185 19 218 143 245 8 49 133 110 81 141 122 234 234 20 45 231 68 180 138 24 56 214 164 17 29 192 178 32 83 185 77 189 36 158 253 73 191 213 7 65 172 108 181 216 113 186 199 28 211 132 68 120 86 37 193 218 26 218 14 161 127 185 41 231 104 223 100 221 228 144 104 206 139 101 204 100 204 69 235 213 11 232 57 43 37 236 98 115 135 118 190 114 89 38 61 2 69 44 217 8 49 0 34 71 96 230 61 178 218 245 104 63 56 235 86 101 2 129 129 0 161 82 133 96 105 149 109 108 4 124 57 242 209 84 23 109 111 87 42 147 204 96 59 253 59 46 172 171 59 27 144 241 252 117 200 63 28 6 238 201 215 73 159 13 76 116 172 69 6 223 14 59 20 9 19 33 227 42 145 9 91 98 47 254 90 60 208 87 2 109 41 56 176 141 179 94 126 134 25 43 215 180 170 223 57 174 144 113 31 27 39 67 145 232 236 108 254 67 19 85 16 231 71 125 205 55 239 162 249 185 133 112 121 156 46 89 135 47 192 17 176 211 59 179 189 17 27 69 2 129 128 155 98 249 133 35 86 130 20 51 203 186 161 196 215 246 130 33 112 194 1 190 39 4 253 84 142 93 173 156 19 11 17 130 232 74 156 226 242 239 192 81 104 158 200 164 188 187 55 240 91 153 134 74 98 210 243 14 66 130 107 112 95 220 83 93 103 113 157 173 81 27 152 236 115 77 76 27 37 55 255 83 139 137 252 89 252 94 144 71 250 149 101 91 235 28 61 14 56 90 18 221 117 176 103 28 8 206 222 221 3 146 137 171 247 104 84 122 90 142 79 109 171 202 41 85 172 189 205 2 129 129 0 77 210 90 188 159 105 138 233 113 142 192 153 120 129 140 82 66 148 67 237 85 220 232 248 240 217 135 10 86 126 5 233 8 5 179 130 164 70 182 77 227 14 207 158 58 144 44 211 251 160 136 50 129 194 118 121 100 126 138 136 35 212 92 187 152 55 117 133 47 251 199 86 197 130 191 188 40 0 154 205 226 71 106 57 132 60 254 125 84 30 157 42 48 52 15 39 65 125 106 105 171 152 166 155 235 19 139 115 144 16 125 83 137 51 87 96 32 19 73 229 201 98 207 114 106 160 70 74 160 13 48 11 6 3 85 29 15 49 4 3 2 0 16
```
I suppose that because of this I can not read data that previously have been encrypted using an asymmetric key.

**Comment by AArnott** [comment]
This might not actually be an issue. Can you confirm that
encryption/decryption is broken? The key length deviation is likely due to
a leading or trailing zero, which might not be significant to actual use.

**Comment by LKuzma** [comment]
Ok. When I add it to my example:
```c#
var encryptedBytes = WinRTCrypto.CryptographicEngine.Encrypt(asymKey, new byte[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 });
System.Diagnostics.Debug.WriteLine("encrypted by asymKey:");
foreach (var item in encryptedBytes)
    sbuild.Append(item + " ");
System.Diagnostics.Debug.WriteLine(sbuild.ToString());
sbuild.Clear();

var encryptedBytesByImportedKey = WinRTCrypto.CryptographicEngine.Encrypt(imported, new byte[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 });
System.Diagnostics.Debug.WriteLine("encrypted by imported key:");
foreach (var item in encryptedBytesByImportedKey)
    sbuild.Append(item + " ");
System.Diagnostics.Debug.WriteLine(sbuild.ToString());
sbuild.Clear();
```
Getting a
```
Java.Lang.RuntimeException: input too large for RSA cipher.
```
when I want to encrypt using the imported Key.
Result:
```
encrypted by asymKey:
65 61 64 38 67 190 10 18 194 27 214 226 84 171 163 31 188 58 228 106 36 63 140 234 191 20 22 212 205 96 65 19 167 187 39 104 128 9 253 123 132 5 112 148 240 173 92 219 49 235 120 166 207 202 10 120 79 192 125 182 4 57 159 166 84 111 120 174 224 3 200 113 220 89 155 86 67 17 104 176 18 240 43 254 169 97 142 193 45 180 52 188 28 42 243 132 54 184 157 171 230 189 230 141 107 231 45 199 180 129 108 70 227 192 126 89 37 187 148 248 22 206 122 156 109 164 96 44 86 138 114 147 233 167 209 163 192 125 185 45 92 68 208 172 42 234 237 44 15 138 196 141 94 8 24 196 163 74 202 185 213 65 129 99 107 51 116 44 26 38 167 33 37 124 141 201 184 228 175 164 209 151 234 71 26 32 118 23 65 239 230 40 30 125 186 195 182 151 139 216 79 96 173 141 84 221 187 161 11 206 3 223 136 89 209 0 112 249 2 160 163 229 230 78 66 22 182 149 132 60 222 213 184 16 155 238 218 205 204 229 107 5 72 174 175 234 234 48 138 215 75 226 70 45 110 32
Unhandled Exception:
Java.Lang.RuntimeException: input too large for RSA cipher.
```
Prevents me to decrypt and encrypt the information using imported key. I noticed it on Android 5.0 and 4.0.3. The same code run on 4.1.1 and 4.4.1 produces the result:
```
KeySize before export: 2048
imported data:
48 130 4 205 2 1 0 48 13 6 9 42 134 72 134 247 13 1 1 1 5 0 4 130 4 168 48 130 4 164 2 1 0 2 130 1 1 0 148 112 188 116 255 185 82 237 144 62 77 151 201 89 160 156 198 235 26 164 183 104 140 143 174 15 232 192 97 7 208 223 48 144 176 152 81 75 195 205 209 28 226 113 165 115 49 248 138 211 26 35 189 212 162 117 74 63 116 201 225 60 107 253 188 68 75 19 27 80 40 106 7 213 19 164 150 216 168 163 46 55 99 85 222 12 25 237 79 47 68 185 246 236 12 71 140 239 112 251 22 233 81 8 99 131 253 3 147 245 228 171 53 60 39 42 12 247 60 109 239 74 217 206 169 19 77 216 241 53 92 9 126 49 216 143 251 124 150 46 63 243 229 32 73 128 104 51 35 7 142 110 100 140 228 107 184 30 53 80 236 83 45 86 211 97 220 34 243 31 149 78 199 117 87 91 155 232 37 131 140 30 133 248 146 198 233 1 226 38 34 170 148 83 165 150 222 13 209 194 229 219 231 156 195 155 232 245 214 53 227 105 237 116 150 95 226 133 67 191 90 8 34 235 26 59 107 220 26 72 84 193 243 240 125 16 215 216 127 62 176 196 46 34 179 243 236 130 33 12 178 35 3 251 19 57 2 3 1 0 1 2 130 1 0 136 72 193 30 2 51 26 42 54 134 157 34 192 160 99 79 205 105 201 58 117 145 173 139 180 217 178 109 254 39 248 227 17 105 166 89 18 58 167 120 237 134 193 120 105 216 203 134 79 158 229 158 239 8 172 103 87 85 207 90 18 75 49 50 242 137 25 12 95 197 70 117 162 223 68 123 150 107 48 217 48 243 117 102 62 201 137 124 81 121 42 118 58 120 26 6 15 37 62 235 168 235 105 148 116 25 152 80 146 165 237 222 42 176 33 156 255 238 118 50 208 194 62 26 249 44 90 9 111 229 195 240 98 225 26 109 70 246 16 2 10 192 154 40 149 115 137 42 198 38 30 216 224 13 24 176 193 92 107 189 77 9 147 181 123 252 37 16 101 223 59 160 62 157 165 190 171 141 86 14 111 151 42 14 64 178 237 124 222 227 34 22 93 195 75 202 158 137 182 57 223 246 185 239 52 194 220 115 219 16 87 165 75 170 151 92 59 189 68 114 206 128 141 189 142 55 74 157 107 213 70 48 210 235 230 101 76 57 102 224 74 167 165 75 156 75 149 184 140 12 32 190 196 238 91 161 2 129 129 0 196 76 132 134 189 136 149 147 123 212 171 253 89 205 125 33 60 206 205 20 155 123 91 163 176 36 65 113 148 199 175 17 240 143 109 92 120 251 68 227 179 117 185 224 140 67 3 121 125 163 116 115 220 200 82 251 81 157 37 208 146 91 94 240 154 225 33 123 57 149 137 202 13 13 18 37 71 177 212 164 99 6 98 188 255 98 77 109 124 181 122 84 209 204 54 236 115 33 177 84 110 243 220 15 241 235 21 12 100 194 239 119 84 94 11 78 238 161 65 148 119 71 26 202 44 112 61 213 2 129 129 0 193 150 10 14 13 185 14 199 214 113 189 175 244 156 235 56 114 42 111 45 34 92 84 108 174 183 88 2 239 181 32 17 134 28 197 46 90 179 60 200 127 156 164 251 119 95 209 2 147 77 221 194 221 121 181 173 41 154 197 206 73 203 4 4 10 17 57 53 204 100 202 177 72 149 34 28 14 215 197 87 76 63 27 4 113 22 163 38 78 203 220 252 174 187 134 111 202 163 210 64 42 18 103 223 71 222 201 45 218 80 173 213 170 188 50 211 230 113 185 237 248 113 252 21 123 252 157 213 2 129 129 0 26 175 64 48 196 225 37 92 56 48 47 245 47 217 4 82 153 94 143 120 102 174 125 69 162 51 59 134 66 48 102 247 44 89 131 7 88 22 232 69 127 230 216 14 211 98 100 189 134 59 203 31 216 250 142 117 69 161 244 59 7 58 78 140 172 177 64 8 237 74 11 194 32 198 15 26 108 69 155 95 129 106 163 207 135 116 8 140 130 176 116 225 196 231 90 236 7 182 163 209 147 135 168 249 173 236 68 31 60 47 87 184 68 10 200 196 193 33 192 181 123 5 122 121 211 82 170 197 2 129 128 190 206 120 82 169 204 1 121 143 48 196 218 173 172 13 92 177 138 137 210 94 207 202 150 229 14 233 188 219 48 23 207 173 244 165 149 174 232 131 188 203 171 70 34 101 227 23 76 6 140 245 75 160 168 218 29 55 168 201 215 46 124 224 20 107 39 214 207 228 10 100 142 161 15 253 114 210 94 114 176 61 2 230 74 86 125 13 138 48 152 87 212 231 21 6 201 33 94 159 89 135 30 101 43 122 215 104 75 189 167 79 240 14 61 57 140 10 211 133 131 189 33 252 105 224 175 121 145 2 129 129 0 124 223 6 62 253 238 192 98 25 248 122 67 181 235 64 147 13 158 251 99 50 116 96 34 192 113 193 243 195 232 64 204 132 105 165 87 134 222 235 206 203 153 29 102 81 52 68 119 88 164 68 84 195 95 170 83 146 40 139 95 136 223 182 205 38 78 169 142 171 248 8 223 61 99 176 161 131 119 93 255 240 61 52 223 145 30 116 26 10 117 227 141 108 252 170 81 161 115 105 106 36 73 201 2 99 217 72 141 194 135 150 49 199 185 91 140 56 235 49 6 161 62 239 75 241 96 253 5 160 13 48 11 6 3 85 29 15 49 4 3 2 0 16 

KeySize after import: 2048
imported data:
48 130 4 205 2 1 0 48 13 6 9 42 134 72 134 247 13 1 1 1 5 0 4 130 4 168 48 130 4 164 2 1 0 2 130 1 1 0 148 112 188 116 255 185 82 237 144 62 77 151 201 89 160 156 198 235 26 164 183 104 140 143 174 15 232 192 97 7 208 223 48 144 176 152 81 75 195 205 209 28 226 113 165 115 49 248 138 211 26 35 189 212 162 117 74 63 116 201 225 60 107 253 188 68 75 19 27 80 40 106 7 213 19 164 150 216 168 163 46 55 99 85 222 12 25 237 79 47 68 185 246 236 12 71 140 239 112 251 22 233 81 8 99 131 253 3 147 245 228 171 53 60 39 42 12 247 60 109 239 74 217 206 169 19 77 216 241 53 92 9 126 49 216 143 251 124 150 46 63 243 229 32 73 128 104 51 35 7 142 110 100 140 228 107 184 30 53 80 236 83 45 86 211 97 220 34 243 31 149 78 199 117 87 91 155 232 37 131 140 30 133 248 146 198 233 1 226 38 34 170 148 83 165 150 222 13 209 194 229 219 231 156 195 155 232 245 214 53 227 105 237 116 150 95 226 133 67 191 90 8 34 235 26 59 107 220 26 72 84 193 243 240 125 16 215 216 127 62 176 196 46 34 179 243 236 130 33 12 178 35 3 251 19 57 2 3 1 0 1 2 130 1 0 136 72 193 30 2 51 26 42 54 134 157 34 192 160 99 79 205 105 201 58 117 145 173 139 180 217 178 109 254 39 248 227 17 105 166 89 18 58 167 120 237 134 193 120 105 216 203 134 79 158 229 158 239 8 172 103 87 85 207 90 18 75 49 50 242 137 25 12 95 197 70 117 162 223 68 123 150 107 48 217 48 243 117 102 62 201 137 124 81 121 42 118 58 120 26 6 15 37 62 235 168 235 105 148 116 25 152 80 146 165 237 222 42 176 33 156 255 238 118 50 208 194 62 26 249 44 90 9 111 229 195 240 98 225 26 109 70 246 16 2 10 192 154 40 149 115 137 42 198 38 30 216 224 13 24 176 193 92 107 189 77 9 147 181 123 252 37 16 101 223 59 160 62 157 165 190 171 141 86 14 111 151 42 14 64 178 237 124 222 227 34 22 93 195 75 202 158 137 182 57 223 246 185 239 52 194 220 115 219 16 87 165 75 170 151 92 59 189 68 114 206 128 141 189 142 55 74 157 107 213 70 48 210 235 230 101 76 57 102 224 74 167 165 75 156 75 149 184 140 12 32 190 196 238 91 161 2 129 129 0 196 76 132 134 189 136 149 147 123 212 171 253 89 205 125 33 60 206 205 20 155 123 91 163 176 36 65 113 148 199 175 17 240 143 109 92 120 251 68 227 179 117 185 224 140 67 3 121 125 163 116 115 220 200 82 251 81 157 37 208 146 91 94 240 154 225 33 123 57 149 137 202 13 13 18 37 71 177 212 164 99 6 98 188 255 98 77 109 124 181 122 84 209 204 54 236 115 33 177 84 110 243 220 15 241 235 21 12 100 194 239 119 84 94 11 78 238 161 65 148 119 71 26 202 44 112 61 213 2 129 129 0 193 150 10 14 13 185 14 199 214 113 189 175 244 156 235 56 114 42 111 45 34 92 84 108 174 183 88 2 239 181 32 17 134 28 197 46 90 179 60 200 127 156 164 251 119 95 209 2 147 77 221 194 221 121 181 173 41 154 197 206 73 203 4 4 10 17 57 53 204 100 202 177 72 149 34 28 14 215 197 87 76 63 27 4 113 22 163 38 78 203 220 252 174 187 134 111 202 163 210 64 42 18 103 223 71 222 201 45 218 80 173 213 170 188 50 211 230 113 185 237 248 113 252 21 123 252 157 213 2 129 129 0 26 175 64 48 196 225 37 92 56 48 47 245 47 217 4 82 153 94 143 120 102 174 125 69 162 51 59 134 66 48 102 247 44 89 131 7 88 22 232 69 127 230 216 14 211 98 100 189 134 59 203 31 216 250 142 117 69 161 244 59 7 58 78 140 172 177 64 8 237 74 11 194 32 198 15 26 108 69 155 95 129 106 163 207 135 116 8 140 130 176 116 225 196 231 90 236 7 182 163 209 147 135 168 249 173 236 68 31 60 47 87 184 68 10 200 196 193 33 192 181 123 5 122 121 211 82 170 197 2 129 128 190 206 120 82 169 204 1 121 143 48 196 218 173 172 13 92 177 138 137 210 94 207 202 150 229 14 233 188 219 48 23 207 173 244 165 149 174 232 131 188 203 171 70 34 101 227 23 76 6 140 245 75 160 168 218 29 55 168 201 215 46 124 224 20 107 39 214 207 228 10 100 142 161 15 253 114 210 94 114 176 61 2 230 74 86 125 13 138 48 152 87 212 231 21 6 201 33 94 159 89 135 30 101 43 122 215 104 75 189 167 79 240 14 61 57 140 10 211 133 131 189 33 252 105 224 175 121 145 2 129 129 0 124 223 6 62 253 238 192 98 25 248 122 67 181 235 64 147 13 158 251 99 50 116 96 34 192 113 193 243 195 232 64 204 132 105 165 87 134 222 235 206 203 153 29 102 81 52 68 119 88 164 68 84 195 95 170 83 146 40 139 95 136 223 182 205 38 78 169 142 171 248 8 223 61 99 176 161 131 119 93 255 240 61 52 223 145 30 116 26 10 117 227 141 108 252 170 81 161 115 105 106 36 73 201 2 99 217 72 141 194 135 150 49 199 185 91 140 56 235 49 6 161 62 239 75 241 96 253 5 160 13 48 11 6 3 85 29 15 49 4 3 2 0 16 

encrypted by asymKey:
111 57 204 167 142 34 179 179 68 54 64 192 160 247 13 251 34 254 107 75 225 66 128 203 172 236 150 129 249 180 124 193 195 123 87 183 104 148 62 240 100 103 135 229 92 90 216 6 38 77 210 53 227 74 157 60 236 235 199 14 43 180 11 130 107 155 136 165 195 34 198 177 86 124 79 110 62 91 35 239 169 66 229 5 57 150 204 75 51 83 4 116 183 105 30 255 43 245 102 209 116 158 234 2 88 70 92 49 187 194 230 158 198 218 93 62 28 43 66 246 163 167 180 136 31 233 91 136 174 220 2 251 94 183 184 101 97 117 90 139 111 113 157 19 43 147 50 243 85 86 73 56 42 191 192 80 249 73 148 136 67 114 130 195 40 164 191 66 44 137 243 77 145 240 115 253 73 205 188 159 124 253 84 29 89 201 103 142 78 80 167 237 157 229 119 98 207 160 225 63 192 175 45 142 168 38 3 103 181 24 220 112 225 208 181 176 163 14 114 102 191 227 148 51 155 155 63 111 87 67 212 237 228 9 146 16 77 72 250 85 166 174 134 134 171 141 214 122 166 223 168 227 13 250 163 196 

encrypted by imported key:
109 24 247 58 1 184 42 230 238 20 126 219 160 15 183 249 166 170 88 78 53 12 13 75 19 140 220 45 76 231 52 124 232 94 210 29 247 190 254 42 18 76 74 171 213 209 11 173 8 169 208 158 185 144 147 132 24 76 209 143 218 48 209 133 7 23 126 38 112 142 23 51 90 74 87 54 44 4 225 221 183 97 153 237 154 76 53 157 9 196 228 210 211 63 6 24 120 103 69 104 243 82 97 83 46 137 183 211 255 190 17 105 27 130 211 106 9 2 107 168 67 53 188 117 14 108 44 146 196 6 179 33 142 47 135 103 202 50 189 184 246 129 26 3 34 212 115 249 199 53 101 73 156 116 174 47 194 44 185 48 202 27 217 214 29 93 199 19 144 130 212 79 199 124 35 133 244 194 70 98 158 200 38 177 48 4 61 141 8 26 83 77 229 139 247 72 183 43 210 226 239 126 179 146 247 188 186 140 120 66 251 237 140 14 224 175 45 233 41 192 181 218 49 168 139 217 225 242 94 255 164 12 40 57 98 131 175 79 22 172 214 232 113 247 123 79 216 223 10 49 67 97 35 249 233 236 
```
It Works perfectly also on IOS 8, Win Phone RT and Win RT.
It looks at the problem with Android 4.0.3 and 5.0. But I believe that you are here the only person that can tell if there is a problem with Android, xamarin or crypto.

**Comment by AArnott** [comment]
As the exception message states, you're encrypting too much with your RSA
key. RSA keys should only be used to encrypt symmetric keys, not arbitrary
text. You should encrypt your plaintext with a symmetric key, and then
encrypt the symmetric key with an RSA public key.
On the receiving side, you decrypt the symmetric key with the RSA private
key, then you decrypt the ciphertext with the symmetric key.

**Comment by LKuzma** [comment]
In my example I try to encrypt only that:
```c#
new byte[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 }
```
It's probably not too much. 
Moreover, the example throws an exception only on Android 4.0.3 and 5.0.

**Comment by AArnott** [comment]
Some RSA implementations can tolerate a larger than normally allowed length
of data.
Please consider cutting your data length in half and see what happens,
unless you can cite the RSA spec that prescribes the maximum length of the
data for a given key size and can prove that 128 bits of data is not too
long for a 2048 bit key.

**Comment by LKuzma** [comment]
Ok. I cut half the length of my data and an exception again appeared. It seems to me that the problem lies not in the length of my data, but in "this" what's going on with the key when I call the import method.
Look at the code that I pasted, sample data which I use, do not throws an exception when I have used the key that I generated at the first lines, an exception is reported only when I want to use an imported key.
Is the imported key should work the same after import, as before export, if you use the same data ?
I do not mean that the result of encryption should be the same, but if I use any key several times, it should allow each time for encrypting the data of the same length.
The way it works in the pcl crypto if you do not use the imported key and does not execute the program on Androide 4.0.3 and 5.0. In relation to the rest of the androids tested the difference revealed in key length after import.

Do you want me to suggest that the problem is almost certainly caused by Android API 21 or Xamarin ?

**Comment by AArnott** [comment]
Thanks for the explanation. I didn't notice before that the freshly created key worked with the same inputs as the imported key failed with. 
It's not likely a bug in Xamarin since recent versions of PCLCrypto bypass Xamarin for RSA encryption (since it's so slow). But it could be a difference in the Android implementations. 
I don't know when I'll have time to check it out. But in the meantime, if you try one of the early versions of PCLCrypto it may work better for you.

**Comment by justremotephone** [comment]
I think this could be related with my pull request:

https://github.com/AArnott/PCLCrypto/pull/17

... and it may be fixed with this changes. I had a similar problem importing public keys from a Windows client....

**Comment by LKuzma** [comment]
Works for me as it should. Thank you, you've done a great job.

**Comment by AArnott** [comment]
Closed with merge of #17

**Comment by AArnott** [comment]
This fix is now included in the 1.0.2 release available on NuGet.

**Comment by justremotephone** [comment]
Great, thanks!
